Mathlib/Topology/MetricSpace/Infsep.lean

```diff
@@ -0,0 +1,599 @@
+    {s : ∀ b : β, Set (π b)} {c : ℝ≥0∞} (h : ∀ b, c ≤ einfsep (s b)) :
+    c ≤ einfsep (Set.pi univ s) := by
+  refine' le_einfsep fun x hx y hy hxy => _
+  rw [mem_univ_pi] at hx hy
+  rcases function.ne_iff.mp hxy with ⟨i, hi⟩
+  exact le_trans (le_einfsep_iff.1 (h i) _ (hx _) _ (hy _) hi) (edist_le_pi_edist _ _ i)
+#align set.le_einfsep_pi_of_le Set.le_einfsep_pi_of_le
+
+end PseudoEMetricSpace
+
+section PseudoMetricSpace
+
+variable [PseudoMetricSpace α] {s : Set α}
+
+theorem subsingleton_of_einfsep_eq_top (hs : s.einfsep = ∞) : s.Subsingleton := by
+  rw [einfsep_top] at hs
+  exact fun _ hx _ hy => of_not_not fun hxy => edist_ne_top _ _ (hs _ hx _ hy hxy)
+#align set.subsingleton_of_einfsep_eq_top Set.subsingleton_of_einfsep_eq_top
+
+theorem einfsep_eq_top_iff : s.einfsep = ∞ ↔ s.Subsingleton :=
+  ⟨subsingleton_of_einfsep_eq_top, Subsingleton.einfsep⟩
+#align set.einfsep_eq_top_iff Set.einfsep_eq_top_iff
+
+theorem Nontrivial.einfsep_ne_top (hs : s.Nontrivial) : s.einfsep ≠ ∞ := by
+  contrapose! hs
+  rw [not_nontrivial_iff]
+  exact subsingleton_of_einfsep_eq_top hs
+#align set.nontrivial.einfsep_ne_top Set.Nontrivial.einfsep_ne_top
+
+theorem Nontrivial.einfsep_lt_top (hs : s.Nontrivial) : s.einfsep < ∞ := by
+  rw [lt_top_iff_ne_top]
+  exact hs.einfsep_ne_top
+#align set.nontrivial.einfsep_lt_top Set.Nontrivial.einfsep_lt_top
+
+theorem einfsep_lt_top_iff : s.einfsep < ∞ ↔ s.Nontrivial :=
+  ⟨nontrivial_of_einfsep_lt_top, Nontrivial.einfsep_lt_top⟩
+#align set.einfsep_lt_top_iff Set.einfsep_lt_top_iff
+
+theorem einfsep_ne_top_iff : s.einfsep ≠ ∞ ↔ s.Nontrivial :=
+  ⟨nontrivial_of_einfsep_ne_top, Nontrivial.einfsep_ne_top⟩
+#align set.einfsep_ne_top_iff Set.einfsep_ne_top_iff
+
+/- ./././Mathport/Syntax/Translate/Basic.lean:635:2: warning: expanding binder collection (x y «expr ∈ » s) -/
+theorem le_einfsep_of_forall_dist_le {d}
+    (h : ∀ (x) (_ : x ∈ s) (y) (_ : y ∈ s) (hxy : x ≠ y), d ≤ dist x y) :
+    ENNReal.ofReal d ≤ s.einfsep :=
+  le_einfsep fun x hx y hy hxy => (edist_dist x y).symm ▸ ENNReal.ofReal_le_ofReal (h x hx y hy hxy)
+#align set.le_einfsep_of_forall_dist_le Set.le_einfsep_of_forall_dist_le
+
+end PseudoMetricSpace
+
+section EMetricSpace
+
+variable [EMetricSpace α] {x y z : α} {s t : Set α} {C : ℝ≥0∞} {sC : Set ℝ≥0∞}
+
+theorem einfsep_pos_of_finite [Finite s] : 0 < s.einfsep := by
+  cases nonempty_fintype s
+  by_cases hs : s.nontrivial
+  · rcases hs.einfsep_exists_of_finite with ⟨x, hx, y, hy, hxy, hxy'⟩
+    exact hxy'.symm ▸ edist_pos.2 hxy
+  · rw [not_nontrivial_iff] at hs
+    exact hs.einfsep.symm ▸ WithTop.zero_lt_top
+#align set.einfsep_pos_of_finite Set.einfsep_pos_of_finite
+
+/- ./././Mathport/Syntax/Translate/Basic.lean:635:2: warning: expanding binder collection (x y «expr ∈ » s) -/
+theorem relatively_discrete_of_finite [Finite s] :
+    ∃ (C : _)(hC : 0 < C), ∀ (x) (_ : x ∈ s) (y) (_ : y ∈ s) (hxy : x ≠ y), C ≤ edist x y := by
```

Comment:
Optional: migrate to `∃ C, 0 < C ∧ _`. We should do this sooner or later and this rarely causes compile errors down the way. Or add a `porting note` about this so that we migrate right after the port.

Also, `∀ x ∈ s, ∀ y ∈ s, x ≠ y → _` is more readable. BTW, we have `Set.Pairwise`. Should we use it here?

Reply:
Feel free to make additional changes. If we want to change `∃` while porting, I suggest that you run `grep -r Mathlib -nie "(. : _)(_ : . ∈ .)"` and fix it in a few other places as well.